### PR TITLE
RES-1228: Modifications to `imagenet_experiment\tune.py` and `restore_utils.py`.

### PR DESCRIPTION
**Pull request by mvacaporale**
Modifications to `imagenet_experiment\tune.py` and `restore_utils.py`:

* Enabled saving of a checkpoint of the initialized model (before this can, at its earliest, be done after 1 epoch of training)
* Added functionality for user defined `train_model` and `evaluate_model` to the `ImagenetExperiment` class. These are now specifiable through the config.
* Enabled param mapping in restore utils. For instance, if a network gets permuted (e.g. swapping the order of ReLU and KWinners), this functionality helps in specifying the mapping of the old names to their new ones. Thus, this enables loading a network from a checkpoint even if the parameters names have slightly changed.

### CODE REVIEW

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py [hunk 1]
```diff
@@ -284,6 +284,9 @@ def _setup(self, config):
             worker_config["rank"] = i
             status.append(w.setup_experiment.remote(worker_config))
 
+        if config.get("checkpoint_at_init", False):
+            self.save()
```
Comment (subutai) [inline nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py]: Which directory does this get saved in?

Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py]: When `save` is called without `checkpoint_dir`, `tune` will place the checkpoint in the results dir postfixed with the "epoch" just like any other checkpoint.

File: nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py [hunk 2]
```diff
@@ -262,7 +262,9 @@ def setup_experiment(self, config):
             multiprocessing.set_start_method("spawn")
 
         # Configure Training data loader
-        self.train_loader = create_train_dataloader(
+        self.create_train_dataloader = config.get(
+            "create_train_dataloader", create_train_dataloader)
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py]: Please add docstring for this new configuration parameter

File: nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py [hunk 3]
```diff
@@ -276,7 +278,9 @@ def setup_experiment(self, config):
         # Configure Validation data loader
         val_dir = config.get("val_dir", "val")
         val_batch_size = config.get("val_batch_size", self.batch_size)
-        self.val_loader = create_validation_dataloader(
+        self.create_validation_dataloader = config.get(
+            "create_validation_dataloader", create_validation_dataloader)
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py]: Please add docstring for this new configuration parameter

File: nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py [hunk 4]
```diff
@@ -299,6 +303,10 @@ def setup_experiment(self, config):
         # Only profile from rank 0
         self.profile = config.get("profile", False) and self.rank == 0
 
+        # Set train and validate methods.
+        self.train_model = config.get("train_model_func", train_model)
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py]: Please add docstring for this new configuration parameter

File: nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py [hunk 5]
```diff
@@ -299,6 +303,10 @@ def setup_experiment(self, config):
...
+        self.evaluate_model = config.get("evaluate_model_func", evaluate_model)
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py]: Please add docstring for this new configuration parameter

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py [hunk 6]
```diff
@@ -284,6 +284,9 @@ def _setup(self, config):
             worker_config["rank"] = i
             status.append(w.setup_experiment.remote(worker_config))
 
+        if config.get("checkpoint_at_init", False):
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py]: Please add docstring for this new configuration parameter

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 1 above)
Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py]: Ray is training the models asynchronously. There is no guarantee the `setup_experiment` call finished at this point.
You should move the `save()` to after `ray.get(status)`

**Comment by mvacaporale** [comment]
@lscheinkman Note: Your requests have been incorporated. Merging this PR

**Pull request merged**